src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -587,36 +637,37 @@ private HeaderStoreItemInfo CloneHeaderInfo(HeaderDescriptor descriptor, HeaderS
             {
                 // We have custom header values. The parsed values are strings.
                 // Custom header values are always stored as string or list of strings.
-                Debug.Assert(sourceInfo.InvalidValue == null, "No invalid values expected for custom headers.");
-                destinationInfo.ParsedValue = CloneStringHeaderInfoValues(sourceInfo.ParsedValue);
+                Debug.Assert(!ExistsInvalidValue(sourceInfo), "No invalid values expected for custom headers.");
+                CloneAndAddValue(destinationInfo, sourceInfo.ParsedAndInvalidValues);
             }
             else
             {
-                // We have a parser, so we also have to copy invalid values and clone parsed values.
+                CloneAndAddValue(destinationInfo, sourceInfo.ParsedAndInvalidValues);
+            }
 
-                // Invalid values are always strings. Strings are immutable. So we only have to clone the
-                // collection (if there is one).
-                destinationInfo.InvalidValue = CloneStringHeaderInfoValues(sourceInfo.InvalidValue);
+            return destinationInfo;
+        }
 
-                // Now clone and add parsed values (if any).
-                if (sourceInfo.ParsedValue != null)
+        private void CloneAndAddValue(HeaderStoreItemInfo destinationInfo, List<HeaderStoreItemInfoValue> source)
+        {
+            destinationInfo.ParsedAndInvalidValues = new List<HeaderStoreItemInfoValue>();
+            foreach (var item in source)
+            {
+                if (item.Value != null)
                 {
-                    List<object>? sourceValues = sourceInfo.ParsedValue as List<object>;
-                    if (sourceValues == null)
-                    {
-                        CloneAndAddValue(destinationInfo, sourceInfo.ParsedValue);
-                    }
-                    else
+                    switch (item.Type)
                     {
-                        foreach (object item in sourceValues)
-                        {
-                            CloneAndAddValue(destinationInfo, item);
-                        }
+                        case HeaderStoreItemInfoValueType.Invalid:
+                            destinationInfo.ParsedAndInvalidValues.Add(new HeaderStoreItemInfoValue { Value = CloneStringHeaderInfoValues(item.Value), Type = HeaderStoreItemInfoValueType.Invalid });
+                            break;
+                        case HeaderStoreItemInfoValueType.Parsed:
+                            CloneAndAddValue(destinationInfo, item.Value);
+                            break;
+                        default:
+                            throw new NotImplementedException(item.Type.ToString());
```

Comment:
I'm not sure (?) that we normally ship code in release builds to protect ourselves against forgetting to update code after we extend an enum some time in the future. I think those cases can be debug asserts.

Reply:
typically it would be 
```c#
switch(.foo)
{
      case X.A:
          ...
         break;
      default:
        Debug.Assert(foo == X.B);
        ...
}
```
That should be shortest.